File: Data Structures & Algorithms/island-perimeter/submission-1.go
var dir = [][2]int{ {-1, 0}, {1, 0}, {0, -1}, {0, 1} }
func islandPerimeter(grid [][]int) int {
    rows, cols := len(grid), len(grid[0])
    var visited = make(map[[2]int]bool)


    for r := 0; r < rows; r++ {
        for c := 0; c < cols; c++ {
            if grid[r][c] == 1 {
                return bfs(r, c, grid, visited)
            }
        }
    }
    return 0
}

func bfs(r, c int, grid [][]int, visited map[[2]int]bool) int {
    queue := [][2]int{{r, c}}
    visited[[2]int{r, c}] = true
    p := 0
    for len(queue) > 0 {
        cell := queue[0]
        queue = queue[1:]

        x, y := cell[0], cell[1]

        for _, d := range dir {
            nx, ny := x + d[0], y + d[1]
            
            if isPerim(nx, ny, grid) {
                p++
            } else if cell := [2]int{nx, ny}; !visited[cell] {
                visited[cell] = true
                queue = append(queue, cell)
            }
        }
    }
    return p
}

func isPerim(r, c int, grid [][]int) bool {
    return r < 0 || c < 0 || r >= len(grid) || c >= len(grid[0]) || grid[r][c] == 0
}

File: Data Structures & Algorithms/islands-and-treasure/submission-1.py
class Solution:
    # get position ofeach chest
    # level by level bfs from the chest and fill each land cell with distance
    # only replace land cell with existing distance
    # if new distance < existing distance
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        def isValid(r, c, grid):
            return r >= 0 and c >= 0 and r < len(grid) and c < len(grid[0])

        DIRS = [[1,0], [-1,0], [0,1], [0,-1]]

        treasure = set()
        for r in range(len(grid)):
            for c in range(len(grid[r])):
                if grid[r][c] == 0:
                    treasure.add((r,c))

        queue = deque() 
        for r, c in treasure:
            queue.append((r, c))
        # level by level bfs
        distance = 1
        while queue:
            size = len(queue)

            # canMove = False
            print(grid)
            for _ in range(len(queue)):
                r, c = queue.popleft()
                for dr, dc in DIRS:
                    nr, nc = dr + r, dc + c
                    if (isValid(nr, nc, grid) and
                    (grid[nr][nc] > 1000 or grid[nr][nc] > grid[r][c])):
                        # canMove = True
                        
                        grid[nr][nc] = distance
                        queue.append((nr, nc))
            
            distance += 1

            


